### PR TITLE
Linux support

### PR DESCRIPTION
**Pull request by muellni**
another approach to get my Linux patches upstream...

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,7 +11,8 @@
 *.pydev*
 *.vcproj
 .idea/
-uid.dll
+lib/uid
+lib/uid.exe
```
Comment (micheljung) [inline .gitignore]: If those files are not checked in, is the developer is required to manually download them?

Comment (muellni) [reply .gitignore]: I'd say the developer should download them or a script or a binary-subrepository or something, but it's not source code...
Of course it has the drawback of not working straight away.

**Comment by duk3luk3** [comment]
The normal approach to include the uid binary would be to add the uid repo as a git submodule and include it in the build process.

**Comment by duk3luk3** [comment]
Please put proper install instructions into the readme. here's what I've tried that doesn't work:

```
virtualenv2 fafclient
cd fafclient
. ./bin/activate
pip install -r ~/git/faf/client/requirements.txt
pip install -e ~/git/faf/client
```

Gives this error:

```
(fafclient) luke@Max fafclient % pip install -e ~/git/faf/client                             ~/scratch/venvs/fafclient
Obtaining file:///home/luke/git/faf/client
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/luke/git/faf/client/setup.py", line 4, in <module>
        import sip
    ImportError: No module named sip

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /home/luke/git/faf/client/
```

**Comment by duk3luk3** [comment]
If I bodge in the Qt dependencies the install works, but that's terribly ugly and probably doesn't actually work...

```
cd lib/python2.7/site-packages
ln -s /usr/lib/python2.7/site-packages/sip.py
ln -s /usr/lib/python2.7/site-packages/sip.so
ln -s /usr/lib/python2.7/site-packages/PyQt4
cd ../../..
pip install -e ~/git/faf/client
Obtaining file:///home/luke/git/faf/client
Installing collected packages: Forged-Alliance-Forever
  Running setup.py develop for Forged-Alliance-Forever
Successfully installed Forged-Alliance-Forever
```

Then I try to run:

```
cd ~/scratch/venvs/fafclient
. ./bin/activate
cd ~/git/faf/client
python2 src
```

It gives me only a dialog saying "An error has occurred in the FAF client. It has been automatically reported."

EDIT: It should be logging errors to somewhere but I have no idea where the logs end up.

**Comment by muellni** [comment]
The log files are located at ~/FAForever/logs.
For the virtualenv, you might use `virtualenv2 --system-site-packages` to have access to your system's pyqt. But I never tried installing the fafclient in a venv. I either install it as a system package or run it from source. I usually use a virtualenv for running it from source, similar to the one you created.

**Comment by duk3luk3** [comment]
Yeah, running the client install wasn't actually what I wanted.

Could you please update this branch to rebase it against develop and add some setup instructions to the readme?

**Comment by duk3luk3** [comment]
Here's the log:

```
2016-11-05 12:05:59,466 INFO     root                           FAF version: None Environment: production
2016-11-05 12:05:59,807 INFO     __main__                       >>> --------------------------- Application Launch
2016-11-05 12:06:00,030 INFO     faf.secondaryServer.Statistic  Instantiating secondary server.
2016-11-05 12:06:00,059 INFO     util                           Read common stylesheet: client/client.css
2016-11-05 12:06:00,064 ERROR    __main__                       Uncaught exception
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/luke/git/faf/client/src/__main__.py", line 122, in <module>
    runFAF()
  File "/home/luke/git/faf/client/src/__main__.py", line 76, in runFAF
    import client
  File "src/client/__init__.py", line 45, in <module>
    instance = ClientWindow()
  File "src/client/_clientwindow.py", line 215, in __init__
    self.setWindowTitle("FA Forever " + util.VERSION_STRING)
TypeError: cannot concatenate 'str' and 'NoneType' objects
```

So, whatever determines util.VERSION_STRING needs to be extended to give a commit hash or a fallback "develop" string.

**Comment by duk3luk3** [comment]
Good catch on the version check, I just checked it out myself and was going to tell you what the problem seemed to be. :+1:

**Comment by duk3luk3** [comment]
This works to launch the client for me now, but login doesn't work - no idea if that's related.

**Comment by duk3luk3** [comment]
@muellni I'll merge this as soon as you add readme update ;)

**Comment by muellni** [comment]
@duk3luk3 what do you think about readme update? 
Linux info is now pretty scattered between wiki and readme.

**Comment by duk3luk3** [comment]
looks good!

Does running tests work for you? I get this:

```
(fafclient) luke@Max client(linux_support_2|✔) % python2 runtests.py
INTERNALERROR> Traceback (most recent call last):                   
INTERNALERROR>   File "_pytest.main", line 80, in wrap_session
INTERNALERROR>     config.do_configure()
INTERNALERROR>   File "_pytest.config", line 618, in do_configure
INTERNALERROR>     self.hook.pytest_configure(config=self)
INTERNALERROR>   File "_pytest.core", line 413, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "_pytest.core", line 424, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "_pytest.core", line 315, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "/home/luke/scratch/venvs/fafclient/lib/python2.7/site-packages/pytestqt/plugin.py", line 167, in pytest_configure
INTERNALERROR>     qt_api.set_qt_api(config.getini('qt_api'))
INTERNALERROR>   File "/home/luke/scratch/venvs/fafclient/lib/python2.7/site-packages/pytestqt/qt_compat.py", line 75, in set_qt_api
INTERNALERROR>     sip.setapi("QDate", 2)
INTERNALERROR> ValueError: API 'QDate' has already been set to version 1
```

If I patch runtests.py to setup the api versions before doing anything else, I get this:

```
(fafclient) luke@Max client(linux_support_2|✚1) % python2 runtests.py
========================================================= test session starts ==========================================================
platform linux2 -- Python 2.7.12 -- py-1.4.25 -- pytest-2.6.3
PyQt4 4.11.4 -- Qt runtime 4.8.7 -- Qt compiled 4.8.7
plugins: qt, mock, cov
collected 31 items 

tests/fa/test_featured.py 
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "_pytest.main", line 84, in wrap_session
INTERNALERROR>     doit(config, session)
INTERNALERROR>   File "_pytest.main", line 122, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "_pytest.core", line 413, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "_pytest.core", line 424, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "_pytest.core", line 315, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "_pytest.main", line 142, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "_pytest.core", line 413, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "_pytest.core", line 424, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "_pytest.core", line 315, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "_pytest.runner", line 65, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "_pytest.runner", line 75, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "_pytest.runner", line 111, in call_and_report
INTERNALERROR>     report = hook.pytest_runtest_makereport(item=item, call=call)
INTERNALERROR>   File "_pytest.main", line 167, in call_matching_hooks
INTERNALERROR>     return hookmethod.pcall(plugins, **kwargs)
INTERNALERROR>   File "_pytest.core", line 417, in pcall
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "_pytest.core", line 424, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "_pytest.core", line 325, in execute
INTERNALERROR>     fin()
INTERNALERROR>   File "/home/luke/scratch/venvs/fafclient/lib/python2.7/site-packages/pytestqt/logging.py", line 47, in pytest_runtest_makereport
INTERNALERROR>     report = outcome.get_result()
INTERNALERROR> AttributeError: 'NoneType' object has no attribute 'get_result'
```

EDIT: Just running `py.test` seems to mostly work.

**Comment by duk3luk3** [comment]
I think this can be merged - can you please rebase it into a smaller number of independently working commits (and no reverse commits)

**Comment by muellni** [comment]
@duk3luk3 ok, i'll try. I started with #448

**Comment by duk3luk3** [comment]
This looks good to me, and together with #453 makes the client work natively in linux.

@Sheeo / @aeoncleanse / someone sign off on this please?

**Comment by duk3luk3** [comment]
@muellni please fix: https://travis-ci.org/FAForever/client/builds/173734962 :-)

**Comment by muellni** [comment]
@duk3luk3 is that the travis test from this PR?

**Comment by muellni** [comment]
ok, I think I found the corresponding source tree, but I couldn't reproduce the error travis is having. I did:
`git clone -b linux-merge-tmp https://github.com/FAForever/client faf-client-test`
and ran the client locally. `call_git_describe()` returns `'0.11.60-81-g0f8b0da'` on my machine, while it looks inside travis this line is throwing and out-of-bounds exception: `https://github.com/FAForever/client/blob/linux-merge-tmp/src/config/version.py#L39`

**Comment by muellni** [comment]
@duk3luk3 It is definitely a problem with travis' clone command `--depth=50` detaches it's current commit from the last tag `0.11.60`, so that `git describe --tags` fails. Omitting `--depth=50` works on my machine.

**Comment by duk3luk3** [comment]
I'm working on it!

**Comment by muellni** [comment]
sorry, I think I was faster

**Comment by duk3luk3** [comment]
To make version-readout work in travis, we either need to change the `--depth 50` parameter to the travis git checkout so a tag can always be found, or change the `git describe` call to add `--always` like in https://github.com/FAForever/client/commit/ee687a9ceaabecc8096ec7f06e046925ecace6c5

Which is preferable? @Sheeo

**Comment by Sheeo** [comment]
I have no preference.

**Comment by duk3luk3** [comment]
OK, I'll go with '--always' then. It shouldn't impact anything other than the travis build.

**Comment by duk3luk3** [comment]
Merged in ee687a9

**Pull request closed without merging**